File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int ans = 0;
    int func(TreeNode* root){
        if(root == nullptr){
            return 0;
        }
        int leftSum = 0, rightSum=0;
        if(root->left){
            leftSum = 1 + func(root->left);
        }
        if(root->right){
            rightSum = 1 + func(root->right);
        }
        ans = max(ans,leftSum+ rightSum);
        return max(leftSum,rightSum);
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int temp = func(root);
        return ans;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def solve(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        left = right = 0
        if root.left:
            left = 1 + self.solve(root.left)

        if root.right:
            right = 1 + self.solve(root.right)

        self.ans = max(self.ans , left + right)

        return max(left,right)
        
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.ans = 0
        self.solve(root)
        return self.ans